### PR TITLE
Add liveliness and lease duration to pub and echo

### PR DESCRIPTION
**Pull request by NourS-d**
Closes Issue #614

### CODE REVIEW

**Comment by NourS-d** [comment]
Pub
```
$ ros2 topic pub /demo/test_demo geometry_msgs/Twist '{linear: {x: -0.0}}' --qos-liveliness-lease-duration 20e9
```

Echo
```
$ ros2 topic echo /demo/test_demo geometry_msgs/Twist --qos-liveliness-lease-duration 20e9 
1630525300.053282 [0]       ros2: using network interface wlp2s0 (udp/192.168.43.167) selected arbitrarily from: wlp2s0, docker0
linear:
  x: -0.0
  y: 0.0
  z: 0.0
angular:
  x: 0.0
  y: 0.0
  z: 0.0
---
$ ros2 topic echo /demo/test_demo geometry_msgs/Twist --qos-liveliness-lease-duration 30e9 
1630525306.729698 [0]       ros2: using network interface wlp2s0 (udp/192.168.43.167) selected arbitrarily from: wlp2s0, docker0
linear:
  x: -0.0
  y: 0.0
  z: 0.0
angular:
  x: 0.0
  y: 0.0
  z: 0.0
---
$ ros2 topic echo /demo/test_demo geometry_msgs/Twist --qos-liveliness-lease-duration 10e9 
1630525312.813764 [0]       ros2: using network interface wlp2s0 (udp/192.168.43.167) selected arbitrarily from: wlp2s0, docker0
[WARN] [1630525313.416047065] [_ros2cli_121115]: New publisher discovered on topic '/demo/test_demo', offering incompatible QoS. No messages will be received from it. Last incompatible policy: LIVELINESS

```

**Comment by NourS-d** [comment]
@audrow Hello, I created a PR for this since it looked relatively straightforward.
https://github.com/ros2/ros2cli/issues/614#issuecomment-907846299
Do you believe some special reasoning is needed for liveliness?

File: ros2topic/ros2topic/verb/echo.py [hunk 1]
```diff
@@ -82,6 +83,15 @@ def add_arguments(self, parser, cli_name):
             help='Quality of service durability setting to subscribe with '
                  '(overrides durability value of --qos-profile option, default: '
                  'Automatically match existing publishers )')
+        parser.add_argument(
+            '--qos-liveliness',
+            choices=rclpy.qos.QoSDurabilityPolicy.short_keys(),
+            help='Quality of service liveliness setting to publish with '
```
Comment (fujitatomoya) [inline ros2topic/ros2topic/verb/echo.py:89]:
```suggestion
            help='Quality of service liveliness setting to subscribe with '
```

**Review by fujitatomoya — CHANGES_REQUESTED** [review]
just one nit, could you fix it?

File: ros2topic/ros2topic/api/__init__.py [hunk 2]
```diff
@@ -168,6 +190,11 @@ def profile_configure_short_keys(
         profile.durability = rclpy.qos.QoSDurabilityPolicy.get_from_short_key(durability)
     if reliability:
         profile.reliability = rclpy.qos.QoSReliabilityPolicy.get_from_short_key(reliability)
+    if liveliness:
+        profile.liveliness = rclpy.qos.QoSLivelinessPolicy.get_from_short_key(
+            liveliness)
```
Comment (audrow) [inline ros2topic/ros2topic/api/__init__.py:195]: Maybe put this on one line? I think the above two lines are longer.

File: ros2topic/test/test_qos_conversions.py [hunk 3]
```diff
@@ -14,14 +14,18 @@
...
 
+from ros2topic.api import duration_ns
 from ros2topic.api import qos_profile_from_short_keys
 
 
 def test_profile_conversion():
     profile = qos_profile_from_short_keys(
         'sensor_data', reliability='reliable', durability='transient_local',
-        depth=10, history='keep_last')
+        depth=10, history='keep_last', liveliness='automatic',
```
Comment (audrow) [inline ros2topic/test/test_qos_conversions.py:24]: Could you test some liveliness options other than automatic, too?

**Review by audrow — CHANGES_REQUESTED** [review]
Just a couple nitpicks - thanks for the PR!

**Comment by audrow** [comment]
@ros-pull-request-builder retest this please.

**Comment by audrow** [comment]
> Do you believe some special reasoning is needed for liveliness?

I'm not sure. @ivanpauno or @jacobperron, any idea?

**Review by fujitatomoya — APPROVED** [review]

File: ros2topic/ros2topic/api/__init__.py [hunk 4]
```diff
@@ -37,6 +38,26 @@ def unsigned_int(string):
     return value
 
 
+def nonnegative_int(inval):
+    ret = int(inval)
+    if ret < 0:
+        # The error message here gets completely swallowed by argparse
```
Comment (ivanpauno) [inline ros2topic/ros2topic/api/__init__.py:44]: maybe, use `ArgumentTypeError` instead

File: ros2topic/ros2topic/api/__init__.py [hunk 5]
```diff
@@ -37,6 +38,26 @@ def unsigned_int(string):
...
+        raise ValueError('Value must be positive or zero')
+    return ret
+
+
+def positive_float(inval):
+    ret = float(inval)
+    if ret <= 0.0:
+        # The error message here gets completely swallowed by argparse
+        raise ValueError('Value must be positive')
```
Comment (ivanpauno) [inline ros2topic/ros2topic/api/__init__.py:53]: same here ArgumentTypeError

File: ros2topic/ros2topic/api/__init__.py [hunk 6]
```diff
@@ -37,6 +38,26 @@ def unsigned_int(string):
     return value
 
 
+def nonnegative_int(inval):
```
Comment (ivanpauno) [inline ros2topic/ros2topic/api/__init__.py:41]: what's the difference with `unsigned_int`?
both are the same

File: ros2topic/ros2topic/api/__init__.py [hunk 7]
```diff
@@ -37,6 +38,26 @@ def unsigned_int(string):
...
+    ret = int(inval)
+    if ret < 0:
+        # The error message here gets completely swallowed by argparse
+        raise ValueError('Value must be positive or zero')
+    return ret
+
+
+def positive_float(inval):
+    ret = float(inval)
```
Comment (ivanpauno) [inline ros2topic/ros2topic/api/__init__.py:50]: If `ValueError` formatting doesn't work well, catching that exception and reraising ArgumentTypeError would be better

File: ros2topic/ros2topic/api/__init__.py (hunk 4 above)
Comment (ivanpauno) [reply ros2topic/ros2topic/api/__init__.py:44]: according to argparse docs, a raised `ValueError` should be formatted correctly, but if that's not the case let's use `ArgumentTypeError`.

File: ros2topic/ros2topic/api/__init__.py [hunk 8]
```diff
@@ -37,6 +38,26 @@ def unsigned_int(string):
...
+    if ret <= 0.0:
+        # The error message here gets completely swallowed by argparse
+        raise ValueError('Value must be positive')
+    return ret
+
+
+def duration_ns(inval):
+    return Duration(nanoseconds=positive_float(inval))
```
Comment (ivanpauno) [inline ros2topic/ros2topic/api/__init__.py:58]: why a float?
this should be an integer, https://github.com/ros2/rclpy/blob/388002a9f235ef5ede0fb76b9fb751d6af0b305f/rclpy/rclpy/duration.py#L23.

**Review by ivanpauno — CHANGES_REQUESTED** [review]
I have a few comments/questions

File: ros2topic/ros2topic/api/__init__.py (hunk 8 above)
Comment (NourS-d) [reply ros2topic/ros2topic/api/__init__.py:58]: True! do you think that nanoseconds is a good choice? should I go for seconds instead

File: ros2topic/ros2topic/api/__init__.py (hunk 6 above)
Comment (NourS-d) [reply ros2topic/ros2topic/api/__init__.py:41]: I just blindly moved things here. It is the same but the unsigned_int has the exception note you mentioned below.

File: ros2topic/ros2topic/verb/echo.py [hunk 9]
```diff
@@ -82,6 +83,15 @@ def add_arguments(self, parser, cli_name):
             help='Quality of service durability setting to subscribe with '
                  '(overrides durability value of --qos-profile option, default: '
                  'Automatically match existing publishers )')
+        parser.add_argument(
+            '--qos-liveliness',
+            choices=rclpy.qos.QoSDurabilityPolicy.short_keys(),
```
Comment (jacobperron) [inline ros2topic/ros2topic/verb/echo.py:88]:
```suggestion
            choices=rclpy.qos.LivelinessPolicy.short_keys(),
```

File: ros2topic/ros2topic/api/__init__.py (hunk 8 above)
Comment (jacobperron) [reply ros2topic/ros2topic/api/__init__.py:58]: I'm indifferent about nanoseconds vs seconds. Maybe the originating of the feature request has a preference (@matthews-jca any thoughts?).

**Review by jacobperron — COMMENTED** [review]
Looks okay to me, with Ivan's comments addressed. One bug fix suggested below.

**Review by ivanpauno — APPROVED** [review]
LGTM with green CI!

**Comment by ivanpauno** [comment]
@ros-pull-request-builder retest this please

**Comment by ivanpauno** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15806/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10484/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/13468/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16086/)

**Comment by ivanpauno** [comment]
there are some new failures in CI, but maybe this just needs a rebase.
@NourS-d could you rebase with master?

If you can also check if you can reproduce the test failures locally that would be great.

**Comment by audrow** [comment]
I rebased - it seems that the same error is present in the failing PR job.

**Comment by fujitatomoya** [comment]
> cannot import name 'nonnegative_int' from 'ros2topic.api' (/tmp/ws/install_isolated/ros2topic/lib/python3.8/site-packages/ros2topic/api/__init__.py)

nonnegative_int has been removed, i think it should be moved to `ros2topic/ros2topic/api/__init__.py`.

**Comment by audrow** [comment]
@NourS-d, would you like to get this in?

**Comment by NourS-d** [comment]
ATM I unfortunately cannot follow up here. However, in two weeks time I can
get into it again. But if you'd like to go ahead :)

**Comment by audrow** [comment]
@mergifyio rebase

**Comment by clalancette** [comment]
We've actually had this functionality for a while, so closing this one out.

**Pull request closed without merging**